aiondemand/AIOD-rest-api
# Update docker-compose

## Change

Update the docker compose file to: 
  - avoid mounting source directories by default, which poses a risk that deployed versions may run code that is not predefined in the image (e.g., local hotfixes instead of through releases)
  - provide a secondary compose file which may be used to mount all the source code instead
  - define the `PYTHONPATH` in the image always, since it's needed to function if there is no mounted source.
  - remove the `version` top level element which is [obsolete](https://docs.docker.com/reference/compose-file/version-and-name/#version-top-level-element-obsolete)
 
 Between the `override.env` introduced in #383, the develop compose file, and the different profiles, running common use case commands becomes very verbose (e.g., `docker compose --env-file=.env --env-file=override.env -f docker-compose.yaml -f docker-compose.dev.yaml --profile openml --profile huggingface-datasets up -d`). I introduced two scripts to provide shorthands:
  - `scripts/up.sh`: takes as arguments any number of profile names and will automatically specify the override environment, as well as the dev compose file if `AIOD_MOUNT` environment variable is set.
  - `scripts/down.sh`: brings down _all_ services (i.e., it uses `--profiles "*"`)


## How to Test

These changes fall outside the scope of unit tests, as they only touch auxiliary scripts and docker.

Since Dockerfiles are modified, images need to be rebuilt:
```bash
docker compose --profile "*" build
./scripts/clean.sh
```
then start the containers, optionally with profiles, e.g.,:
```
./scripts/up.sh openml huggingface-datasets zenodo-datasets examples
```
This should bring up all containers, honoring the overrides in `override.env` but without mounting your local version of the source code.
Modify the local source code, e.g., leave a comment in `src/main.py` and verify it isn't present in a running container
```
docker exec -it apiserver /bin/bash
cat main.py
```

To test the containers with mounted source code, simply set `AIOD_MOUNT=1` in `override.env`. Then re-up the containers, e.g.:
```
./scripts/down.sh
./scripts/up.sh examples
```
When you check the container's content, you should see it now reflects your local changes.

## Checklist
  - No unintended changes have been committed.
  - The changes in isolation seem reasonable.
  - Anything that may be odd or unintuitive is provided with a GitHub comment explaining it (but consider if this should not be a code comment or in the documentation instead).

## Related Issues
